src/backend/cuda/kernel/copy.cuh

```diff
@@ -94,40 +94,82 @@ OTHER_SPECIALIZATIONS(uchar)
 OTHER_SPECIALIZATIONS(char)
 OTHER_SPECIALIZATIONS(common::half)
 
-template<typename inType, typename outType, bool same_dims>
-__global__ void copy(Param<outType> dst, CParam<inType> src,
-                     outType default_value, double factor, const dims_t trgt,
-                     uint blk_x, uint blk_y) {
-    const uint lx = threadIdx.x;
-    const uint ly = threadIdx.y;
-
-    const uint gz         = blockIdx.x / blk_x;
-    const uint gw         = (blockIdx.y + (blockIdx.z * gridDim.y)) / blk_y;
-    const uint blockIdx_x = blockIdx.x - (blk_x)*gz;
-    const uint blockIdx_y =
-        (blockIdx.y + (blockIdx.z * gridDim.y)) - (blk_y)*gw;
-    const uint gx = blockIdx_x * blockDim.x + lx;
-    const uint gy = blockIdx_y * blockDim.y + ly;
-
-    const inType *in = src.ptr + (gw * src.strides[3] + gz * src.strides[2] +
-                                  gy * src.strides[1]);
-    outType *out     = dst.ptr + (gw * dst.strides[3] + gz * dst.strides[2] +
-                              gy * dst.strides[1]);
-
-    int istride0 = src.strides[0];
-    int ostride0 = dst.strides[0];
-
-    if (gy < dst.dims[1] && gz < dst.dims[2] && gw < dst.dims[3]) {
-        int loop_offset = blockDim.x * blk_x;
-        bool cond = gy < trgt.dim[1] && gz < trgt.dim[2] && gw < trgt.dim[3];
-        for (int rep = gx; rep < dst.dims[0]; rep += loop_offset) {
-            outType temp = default_value;
-            if (same_dims || (rep < trgt.dim[0] && cond)) {
-                temp = convertType<inType, outType>(
-                    scale<inType>(in[rep * istride0], factor));
-            }
-            out[rep * ostride0] = temp;
-        }
+template<typename inType, typename outType, bool SAME_DIMS, bool LOOP1,
+         bool LOOP2, bool FACTOR>
+__global__ void reshapeCopy(Param<outType> dst, CParam<inType> src,
```

Comment:
Updated to scaledCopy.

Since the kernel also has the possibility to change formats (float to double, float to int, ...) I had called it reshapeCopy.  The format changes, also happen on portions in the output buffer for the elements outside of the copy.  The total number of elements is indeed never changed.